internal/database/etcd/etcd.go

```diff
@@ -68,12 +68,21 @@ func (etcd Etcd) Open(ctx context.Context, id uuid.UUID) io.ReadWriter {
 	}
 }
 
-// Write writes data to etcd
+// Write writes data to etcd. If data is nil, the current key will be deleted from the database.
 func (etcd Etcd) Write(p []byte) (int, error) {
 	if etcd.ID == uuid.Nil {
 		return 0, errors.New("please create a new etcd client using Open")
 	}
 	key := fmt.Sprintf("%s/%s", etcd.treePrefix, etcd.ID.String())
+
+	if p == nil {
+		_, err := etcd.client.Delete(etcd.ctx, key)
+		if err != nil {
+			return 0, fmt.Errorf("Failed to delete key %s: %s", key, err.Error())
+		}
+		return 0, nil
+	}
```

Comment:
Why is this added? and why in the write function?

Reply:
I think the question is why are we adding delete in the write function. A question I also have

Reply:
If all potential database clients are supposed to conform to io.ReadWriter interface and be swappable, we have no other choice.

The io.ReadWriter interface seems to be more-suitable for a storage like a binary file, but here it requires workarounds to fit it into a key-value pair storage like etcd.

Reply:
I would rather we write a null value into the database rather than having this function do things that it shouldn't do.

If deletion is required then we should have a separate interface for that and in the Stop function cast the database to that interface to see if it is possible to delete.

```
client := h.database.Open(r.Context(), identifier)
c, canDelete := client.(database.Deleter)
if canDelete {
  err := client.Delete()
} else {
  _, err = client.Write(nil)
}
```